Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingObserver.cs

```diff
@@ -225,80 +208,51 @@ public bool SetObserverOrigin(Vector3 origin)
+                || (outstandingMeshRequest.Value.Collider != cookedData.outputCollider)
+                )
             {
-                dataReady(cookedData, outputWritten, elapsedCookTimeSeconds);
-            }
-        }
-
-        private void CleanupSurface(GameObject surface)
-        {
-            // Destroy the meshes, and add the surface back for reuse
-            CleanupMeshes(surface.GetComponent<MeshFilter>().sharedMesh, surface.GetComponent<MeshCollider>().sharedMesh);
-            availableSurfaces.Enqueue(surface);
-            surface.name = "Unused Surface";
-            surface.SetActive(false);
-        }
+                Debug.LogError(string.Format("Got mismatched OnDataReady for surface {0} while request for surface {1} was outstanding.",
+                    cookedData.id.handle,
+                    outstandingMeshRequest.Value.ID
+                    ));
 
-        private void CleanupMeshes(Mesh visualMesh, Mesh colliderMesh)
-        {
-            if (colliderMesh != null && colliderMesh != visualMesh)
-            {
-                Destroy(colliderMesh);
+                ClearOutstandingMeshRequest();
+                return;
             }
 
-            if (visualMesh != null)
+            if (ObserverState != ObserverStates.Running)
             {
-                Destroy(visualMesh);
+                Debug.Log(string.Format("Got OnDataReady for surface {0}, but observer was no longer running.",
+                    cookedData.id.handle
+                    ));
+
+                ClearOutstandingMeshRequest();
+                return;
             }
-        }
 
-        private GameObject GetSurfaceObject(int surfaceID, Transform parentObject)
-        {
-            //If we have surfaces ready for reuse, use those first
-            if (availableSurfaces.Count > 1)
+            if (!outputWritten)
             {
-                GameObject existingSurface = availableSurfaces.Dequeue();
-                existingSurface.SetActive(true);
-                existingSurface.name = string.Format("Surface-{0}", surfaceID);
-
-                UpdateSurfaceObject(existingSurface, surfaceID);
-
-                return existingSurface;
+                ClearOutstandingMeshRequest();
+                return;
             }
 
-            // If we are adding a new surface, construct a GameObject
-            // to represent its state and attach some Mesh-related
-            // components to it.
-            GameObject toReturn = AddSurfaceObject(null, string.Format("Surface-{0}", surfaceID), transform, surfaceID);
-
-            toReturn.AddComponent<WorldAnchor>();
+            outstandingMeshRequest.Value.Renderer.enabled = SpatialMappingManager.Instance.DrawVisualMeshes;
 
-            return toReturn;
+            // TODO: figure out if we need to do something special because Unity supposedly doesn't properly cleanup baked collision data...
```

Comment:
I added explicit mesh destruction back in and it stopped the meshes from leaking.  Huzzah! :)